fontbe/src/glyphs.rs

```diff
@@ -763,71 +814,13 @@ fn compute_composite_bboxes(context: &Context) -> Result<(), Error> {
-        // Hopefully we can figure out some of those bboxes!
-        for composite in composites.iter() {
-            let glyph_name = &composite.name;
-            let RawGlyph::Composite(composite) = &composite.data else {
-                unreachable!("we just checked that these are all composites");
-            };
-
-            let mut missing_boxes = false;
-            let boxes: Vec<_> = composite
-                .components()
-                .iter()
-                .filter_map(|c| {
-                    if missing_boxes {
-                        return None; // can't succeed
-                    }
-                    let ref_glyph_name = glyph_order.glyph_name(c.glyph.to_u16() as usize).unwrap();
-                    let bbox = bbox_acquired.get(ref_glyph_name).copied().or_else(|| {
-                        glyphs
-                            .get(ref_glyph_name)
-                            .map(|g| g.as_ref().clone())
-                            .and_then(|g| match &g.data {
-                                RawGlyph::Composite(..) => None,
-                                RawGlyph::Empty => None,
-                                RawGlyph::Simple(simple_glyph) => Some(bbox2rect(simple_glyph.bbox)),
-                            })
-                    });
-                    if bbox.is_none() {
-                        trace!("Can't compute bbox for {glyph_name} because bbox for {ref_glyph_name} isn't ready yet");
-                        missing_boxes = true;
-                        return None; // maybe next time?
-                    };
-
-                    // The transform we get here has changed because it got turned into F2Dot14 and i16 parts
-                    // We could go get the "real" transform from IR...?
-                    let affine = affine_for(c);
-                    let transformed_box = affine.transform_rect_bbox(bbox.unwrap());
-                    Some(transformed_box)
-                })
-                .collect();
-            if missing_boxes {
-                trace!("bbox for {glyph_name} not yet resolveable");
-                continue;
-            }
-
-            let bbox = boxes.into_iter().reduce(|acc, e| acc.union(e)).unwrap();
-            trace!("bbox for {glyph_name} {bbox:?}");
-            bbox_acquired.insert(glyph_name.clone(), bbox);
-        }
 
-        // Kerplode if we didn't make any progress this spin
-        composites.retain(|composite| !bbox_acquired.contains_key(&composite.name));
-        if pending == composites.len() {
-            return Err(Error::CompositesStalled(
-                composites.iter().map(|g| g.name.clone()).collect(),
-            ));
-        }
+    for (glyph_name, glyph) in glyphs.iter().filter_map(|(gn, g)| match &g.data {
+        RawGlyph::Composite(composite) => Some((*gn, composite)),
+        RawGlyph::Simple(..) | RawGlyph::Empty => None,
+    }) {
+        let bbox = bbox_of_composite(&glyph_order, &glyphs, glyph, Affine::IDENTITY)?;
```

Comment:
you went for a recursive traversal instead of iterative one. though unlikely, should you prevent from a component cycle triggering an infinite recursion? Python has a max recursion depth after which it raises an exception. Does Rust protect you as well?

Reply:
I think for glyphs sources at least we already check against component cycles when we propagate anchors, see glyphs-reader's propagate_anchors.rs (depth_sorted_composite_glyphs)

Reply:
Let me take confirmation of no cycles as a follow-on. I have the vague notion we might already do it but I'll double-check.

Reply:
https://github.com/googlefonts/fontc/issues/1062